src/library/scala/io/BufferedSource.scala

```diff
@@ -83,17 +84,20 @@ class BufferedSource(inputStream: InputStream, bufferSize: Int)(implicit val cod
 
   override def getLines(): Iterator[String] = new BufferedLineIterator
 
-  /** Efficiently converts the entire remaining input into a string. */
-  override def mkString = {
-    // Speed up slurping of whole data set in the simplest cases.
-    val allReader = decachedReader
-    val sb = new StringBuilder
-    val buf = new Array[Char](bufferSize)
-    var n = 0
-    while (n != -1) {
-      n = allReader.read(buf)
-      if (n>0) sb.appendAll(buf, 0, n)
-    }
-    sb.result
-  }
+  /** Efficiently appends the entire remaining input. */
+  override def addString(sb: StringBuilder, start: String, sep: String, end: String): StringBuilder =
+    if (sep.isEmpty) {
+      val allReader = decachedReader
+      val buf = new Array[Char](bufferSize)
+      val jsb = sb.underlying
+
+      jsb.append(start)
+      var n = allReader.read(buf)
+      while (n != -1) {
+        jsb.append(buf, 0, n)
+        n = allReader.read(buf)
+      }
+      jsb.append(end)
+      sb
+    } else mkString.addString(sb, start, sep, end)
```

Comment:
This is great for code reuse but not so good for memory usage.  I'd push the logic back into the body of the code.

Reply:
Alternatively, just comment that only the `sep = ""` case is efficient.  The other case is admittedly pretty weird.

Reply:
I don’t think this is bad memory usage. All the chars are going into memory already.

In fact this can be quite compact with java compressed strings...  I doubt you could get better though could probably avoid some array copies.

Reply:
I guess storing the entire string twice is not as bad as it could be due to the ~1.5x expected buffer size compared to the final string size.  For such a rare case I guess it isn't worth optimizing more.

Reply:
see https://github.com/scala/scala/pull/6878#discussion_r202232620

I thought it was quite clever to reuse the StringOps impl.  But I could probably adjust to use the ArrayOps impl instead for slightly better perf.

Reply:
I'll add a comment, you're right that this version does consume more memory temporarily than before.